jetty-http2/http2-http-client-transport/src/main/java/org/eclipse/jetty/http2/client/http/HttpReceiverOverHTTP2.java

```diff
@@ -230,198 +220,68 @@ public void onFailure(Throwable failure, Callback callback)
-        private boolean stalled;
+        private volatile boolean notifySuccess;
 
         private ContentNotifier(HttpReceiverOverHTTP2 receiver)
         {
             this.receiver = receiver;
         }
 
-        private void offer(HttpExchange exchange, DataFrame frame, Callback callback)
-        {
-            DataInfo dataInfo = new DataInfo(exchange, frame, callback);
-            if (LOG.isDebugEnabled())
-                LOG.debug("Queueing content {}", dataInfo);
-            enqueue(dataInfo);
-            process(false);
-        }
-
-        private void enqueue(DataInfo dataInfo)
+        public void receive(Stream stream, HttpExchange exchange)
         {
-            try (AutoLock l = lock.lock())
-            {
-                queue.offer(dataInfo);
-            }
+            if (notifySuccess && ((IStream)stream).dataSize() == 0)
+                receiver.responseSuccess(exchange);
+            else
+                stream.demand(1);
         }
 
-        private void process(boolean resume)
+        private void process(Stream stream, HttpExchange exchange, DataFrame dataFrame, Callback callback)
         {
-            // Allow only one thread at a time.
-            boolean busy = active(resume);
-            if (LOG.isDebugEnabled())
-                LOG.debug("Resuming({}) processing({}) of content", resume, !busy);
-            if (busy)
-                return;
-
-            // Process only if there is demand.
-            try (AutoLock l = lock.lock())
-            {
-                if (!resume && demand() <= 0)
-                {
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("Stalling processing, content available but no demand");
-                    active = false;
-                    stalled = true;
-                    return;
-                }
-            }
-
-            while (true)
+            if (dataFrame.getData().hasRemaining())
             {
-                if (dataInfo != null)
-                {
-                    if (dataInfo.frame.isEndStream())
-                    {
-                        receiver.responseSuccess(dataInfo.exchange);
-                        // Return even if active, as reset() will be called later.
-                        return;
-                    }
-                }
-
-                try (AutoLock l = lock.lock())
-                {
-                    dataInfo = queue.poll();
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("Processing content {}", dataInfo);
-                    if (dataInfo == null)
-                    {
-                        active = false;
-                        return;
-                    }
-                }
-
-                ByteBuffer buffer = dataInfo.frame.getData();
-                Callback callback = dataInfo.callback;
-                if (buffer.hasRemaining())
+                if (dataFrame.isEndStream())
+                    notifySuccess = true;
```

Comment:
The fact that you're flipping the flag to true only when the data frame's buffer is not empty makes me not at ease, even it that has no impact in the current implementation.

Reply:
It only needed there because in the `else` branch there is implicit demand and we don't call the application, so we can notify immediately the `success` event.